java/languages.antlr/src/org/netbeans/modules/languages/antlr/AntlrTokenId.java

```
@@ -0,0 +1,124 @@
+import org.netbeans.api.lexer.LanguagePath;
+import org.netbeans.api.lexer.Token;
+import org.netbeans.api.lexer.TokenId;
+import org.netbeans.modules.languages.antlr.v3.Antlr3Lexer;
+import org.netbeans.modules.languages.antlr.v4.Antlr4Lexer;
+import org.netbeans.spi.lexer.LanguageEmbedding;
+import org.netbeans.spi.lexer.LanguageHierarchy;
+import org.netbeans.spi.lexer.LanguageProvider;
+import org.netbeans.spi.lexer.Lexer;
+import org.netbeans.spi.lexer.LexerRestartInfo;
+import org.openide.filesystems.FileObject;
+import org.openide.util.Lookup;
+
+/**
+ *
+ * @author lkishalmi
+ */
+public enum AntlrTokenId implements TokenId {
+
+    ACTION("action"),
+    COMMENT("comment"),
+    ERROR("error"),
+    KEYWORD("keyword"),
+    NUMBER("number"),
+    PUNCTUATION("punctuation"),
+    REGEXP_CHARS("regexp-chars"),
+    RULE("rule"),
+    STRING("string"),
+    TOKEN("token"),
+    WHITESPACE("whitespace");
+
+    private final String category;
+
+    AntlrTokenId(String category) {
+        this.category = category;
+    }
+
+    @Override
+    public String primaryCategory() {
+        return category;
+    }
+
+    public static final String MIME_TYPE = "text/x-antlr";
+
+    private static final Language<AntlrTokenId> language =
+            new LanguageHierarchy<AntlrTokenId>() {
+
+                @Override
+                protected String mimeType() {
+                    return AntlrTokenId.MIME_TYPE;
+                }
+
+                @Override
+                protected Collection<AntlrTokenId> createTokenIds() {
+                    return EnumSet.allOf(AntlrTokenId.class);
+                }
+
+                @Override
+                protected Lexer<AntlrTokenId> createLexer(LexerRestartInfo<AntlrTokenId> info) {
+                    FileObject fo =(FileObject) info.getAttributeValue(FileObject.class);
+                    if (fo != null) {
+                        return "g4".equals(fo.getExt()) ? new Antlr4Lexer(info) : new Antlr3Lexer(info);
```

I saw this at least twice and I think it is a hack. It will fail hard, if the user declares a new file mapping (for example `.antlr3` => `text/x-antlr3`). In my mind this would be much cleaner, if there would be two mime-types (`text/x-antlr3` and `text/x-antlr4`).

---

Yes, this is the hack behind the one language two grammar problem.